### PR TITLE
Add integration tests for Helix Java APIs

### LINKED ISSUE #891: Test newly instrumented Helix Java APIs
We need a comprehensive integration test that tests the realm-aware functionalities of new Helix Java APIs and fix any bugs.

### PR DESCRIPTION
**Pull request by narendly**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #891

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This commit adds a comprehensive integration test for Helix Java APIs. All Helix Java APIs are tested using regular resource rebalancing and task framework.

There are a few minor changes to Helix Java API code. Those will be a separate PR for easier reviewing. Note that this PR contains all of those changes.

Depends on the following PRs:
https://github.com/apache/helix/pull/893
https://github.com/apache/helix/pull/894
https://github.com/apache/helix/pull/895
https://github.com/apache/helix/pull/896

### Tests

- [x] The following tests are written for this issue:

TestMultiZkEnvironment

- [x] The following is the result of the "mvn test" command on the appropriate module:

```
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestEnableCompression.testEnableCompressionResource:117 expected:<true> but was:<false>
[INFO]
[ERROR] Tests run: 1097, Failures: 1, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:10 h
[INFO] Finished at: 2020-03-16T21:31:06-07:00
[INFO] ------------------------------------------------------------------------
```

mvn test -Dtest=TestControllerLeadershipChange,TestEnableCompression

```
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 31.587 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:12 min
[INFO] Finished at: 2020-03-13T14:02:11-07:00
[INFO] ------------------------------------------------------------------------
```

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation (Optional)

- [ ] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+    // Create a Mock MSDS
+    final String msdsHostName = "localhost";
+    final int msdsPort = 11117;
+    final String msdsNamespace = "multiZkTest";
+    _msds = new MockMetadataStoreDirectoryServer(msdsHostName, msdsPort, msdsNamespace,
+        _rawRoutingData);
+    _msds.startServer();
+
+    // Turn on multiZk mode in System config
+    System.setProperty(SystemPropertyKeys.MULTI_ZK_ENABLED, "true");
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java]: Maybe it is more accurate for us to back up this system property and recover it after the test? If we simply clear the property after class, there may be a case that another test relying on this property. If the property is cleared, the test may fail. I understand the property should be set and cleared in each test class/method. But just in case.

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+    // MSDS endpoint: http://localhost:11117/admin/v2/namespaces/multiZkTest
+    System.setProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY,
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java]: Do we still need to keep this property after class?

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    }, TestHelper.WAIT_DURATION));
+
+    // Tear down zookeepers
+    ZK_SERVER_MAP.forEach((zkAddress, zkServer) -> zkServer.shutdown());
+
+    // Stop MockMSDS
+    _msds.stopServer();
+
+    // Turn off multiZK mode in System config
+    System.clearProperty(SystemPropertyKeys.MULTI_ZK_ENABLED);
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java]: Just a corner case, if there is exception before clearProperty(), this property will be kept and may affect other tests? Usually what I do is using try...finally.

**Comment by pkuwm** [comment]
@narendly Can you also rebase this PR?

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java (hunk 1 above)
Comment (narendly) [reply helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java]: As you mentioned, system configs should be reset for each test but I could certainly add what youre suggesting :)

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java (hunk 2 above)
Comment (narendly) [reply helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java]: Should be removed.

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java (hunk 3 above)
Comment (narendly) [reply helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java]: Again, I can wrap it in try-finally but each test is responsible for controlling its own System property configs. So a well-written test with a proper beforeClass or a setup method shouldn't be affected. But I agree - wrapping it in a try-finally is not a bad idea so I will add it.

File: helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+/**
+ * TestMultiZkEnvironment spins up multiple in-memory ZooKeepers with a pre-configured
+ * cluster-Zk realm routing information.
+ * This test verifies that all Helix Java APIs work as expected.
+ */
+public class TestMultiZkEnvironment {
```
[the remark this answers is not in the document]
Comment (narendly) [reply helix-core/src/test/java/org/apache/helix/integration/multizk/TestMultiZkEnvironment.java]: I updated the name as you suggested. Thanks :)
